python/ray/__init__.py

```diff
@@ -257,5 +260,18 @@ def __getattr__(self, attr):
 ]
 
 
+if sys.version_info < (3, 7):
+    # TODO(Clark): Remove this one we drop Python 3.6 support.
+    from ray import autoscaler  # noqa: F401
+    from ray import data  # noqa: F401
+    from ray import workflow  # noqa: F401
+else:
+    # Delay importing of expensive, isolated subpackages.
+    def __getattr__(name: str):
+        if name in _subpackages:
+            return importlib.import_module("." + name, __name__)
+        raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

Comment:
Note that, similar to an object's `__getattr__` method, this `__getattr__` will only be called if normal attribute looking (e.g. via `module.__getattribute__`) fails, so this will not run for any of the above subpackages that are eagerly imported and should only run for the packages in `_subpackages` (or non-existent module attributes, which will error anyway).